source/modulo_new_core/include/modulo_new_core/communication/PublisherInterface.hpp

```diff
@@ -38,4 +38,28 @@ class PublisherInterface : public std::enable_shared_from_this<PublisherInterfac
   std::shared_ptr<MessagePairInterface> message_pair_;
 };
 
+template<typename PubT, typename MsgT>
+inline std::shared_ptr<PublisherHandler<PubT, MsgT>> PublisherInterface::get_publisher(bool validate_pointer) {
+  std::shared_ptr<PublisherHandler<PubT, MsgT>> publisher_ptr;
+  try {
+    publisher_ptr = std::dynamic_pointer_cast<PublisherHandler<PubT, MsgT>>(this->shared_from_this());
+  } catch (const std::exception& ex) {
+    if (validate_pointer) {
+      throw exceptions::InvalidPointerException("Publisher interface is not managed by a valid pointer");
+    }
+  }
+  if (publisher_ptr == nullptr && validate_pointer) {
+    throw exceptions::InvalidPointerCastException(
+        "Unable to cast publisher interface to a publisher pointer of requested type"
+    );
+  }
+  return publisher_ptr;
+}
+
+static std::shared_ptr<PublisherInterface> create_publisher_interface(const std::shared_ptr<PublisherInterface>& publisher, const std::shared_ptr<MessagePairInterface>& message_pair) {
+  auto pub = std::shared_ptr<PublisherInterface>(publisher);
+  pub->set_message_pair(message_pair);
```

Comment:
These lines are a bit weird. If you want to make a copy of the argument pointer to reference the same object, it would suffice to write `auto pub = publisher;` I understand the intention is to pass a pointer to a PublisherHandler as the arg, but since the argument type is declared in the function signature it doesn't really need to be duplicated. 

But, moreover, the `publisher` pointer that is passed to the function will be copy constructed anyway if a pointer to a derived type pointer (e.g. PublisherHandler) is passed, so another copy is not really useful. Finally keep in mind that a const pointer is not the same as a pointer to const. All the pointer copies still refer to the non-const object which can be modified through the pointer. 

So, you can just do:
```suggestion
  publisher->set_message_pair(message_pair);
```

Reply:
Hm yeah that makes sense, thanks